stdlib/prometheus/from.go

```diff
@@ -0,0 +1,587 @@
+	}
+
+	return fromNode, true, nil
+}
+func createFromOpSpec(args flux.Arguments, administration *flux.Administration) (flux.OperationSpec, error) {
+	spec := &FromOpSpec{}
+
+	spec.Matcher = make([]*prompb.LabelMatcher, 0)
+
+	if matcher, ok, err := args.GetString("name"); err != nil {
+		return nil, err
+	} else if ok {
+		nameMatcher := &prompb.LabelMatcher{Type: prompb.LabelMatcher_EQ, Name: "__name__", Value: matcher}
+		spec.Matcher = append(spec.Matcher, nameMatcher)
+	}
+
+	if url, ok, err := args.GetString("url"); err != nil {
+		return nil, err
+	} else if ok {
+		if url == "" {
+			return nil, fmt.Errorf("Invalid PromQl url in %q", FromKind)
+		}
+		spec.URL = url
+	}
+
+	if query, ok, err := args.GetString("query"); err != nil {
+		return nil, err
+	} else if ok {
+		if query == "" {
+			return nil, fmt.Errorf("Invalid PromQl query in %q", FromKind)
+		}
+		spec.HasQuery = true
+		spec.Query = query
+	}
+	if user, ok, err := args.GetString("user"); err != nil {
+		return nil, err
+	} else if ok {
+		spec.User = user
+		spec.HasAuth = true
+	}
+	if password, ok, err := args.GetString("password"); err != nil {
+		return nil, err
+	} else if ok {
+		spec.Password = password
+		spec.HasAuth = true
+	}
+	if step, has, err := args.GetDuration("step"); err != nil {
+		return nil, err
+	} else if has {
+		spec.Step = time.Duration(step)
+	} else {
+		spec.Step = defaultStep
+	}
+	return spec, nil
+}
+
+func newFromOp() flux.OperationSpec {
+	return new(FromOpSpec)
+}
+
+func (s *FromOpSpec) Kind() flux.OperationKind {
+	return FromKind
+}
+
+type FromProcedureSpec struct {
+	plan.DefaultCost
+	URL       string
+	Query     string
+	User      string
+	Password  string
+	Matcher   []*prompb.LabelMatcher
+	Step      time.Duration
+	hasQuery  bool
+	hasAuth   bool
+	Bounds    flux.Bounds
+	BoundsSet bool
+	FilterSet bool
+	Start     time.Time
+	End       time.Time
+}
+
+func newFromProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*FromOpSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+	return &FromProcedureSpec{
+		URL:      spec.URL,
+		Query:    spec.Query,
+		User:     spec.User,
+		Password: spec.Password,
+		Matcher:  spec.Matcher,
+		Step:     spec.Step,
+		hasQuery: spec.HasQuery,
+		hasAuth:  spec.HasAuth,
+	}, nil
+}
+
+func (s *FromProcedureSpec) TimeBounds(predecessorBounds *plan.Bounds) *plan.Bounds {
+	if s.BoundsSet {
+		bounds := &plan.Bounds{
+			Start: values.ConvertTime(s.Bounds.Start.Time(s.Bounds.Now)),
+			Stop:  values.ConvertTime(s.Bounds.Stop.Time(s.Bounds.Now)),
+		}
+		return bounds
+	}
+	return nil
+}
+
+func (s *FromProcedureSpec) Kind() plan.ProcedureKind {
+	return FromKind
+}
+func (s *FromProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(FromProcedureSpec)
+	ns.URL = s.URL
+	ns.User = s.User
+	ns.Password = s.Password
+	ns.BoundsSet = s.BoundsSet
+	ns.Bounds = s.Bounds
+	ns.Query = s.Query
+	ns.Matcher = s.Matcher
```

Comment:
This copy needs to be a deep copy, the slice needs to be copied and its contents copied into the new slice.